digitalfabrik/entitlementcard
# 1666: Immutable state for cards and extensions

### Short description


As discussed in #1666, mutating react state is bad practice and commonly leads to bugs as react is not aware of any changes.
Furthermore, mutating objects in place is generally not advisable as it introduces side effects and is also a source of bugs and makes the code less understandable.

This PR aims to address this issues in the case of cards and extensions by splitting the pure state (to be used in useStates/hooks) and helper methods around it to construct/mutate/check this state.

### Proposed changes



- Use simple objects for extensions
- Use correct react components for extension forms
- Use simple objects for cards (blueprints)
- Refactoring helper functions to reflect these changes

### Resulting Improvements

- No more `notifyUpdate` necessary
- Simple and easy adjustment of state (typesafe)
- Easy accessing of extension values

### Side effects


Hopefully none?

### Testing
1. Single card creation for bavaria, nuernberg and koblenz (/erstellen)
2. Create cards from link with queryParams find examples here: https://github.com/digitalfabrik/entitlementcard/pull/1089 Please also check correct addressData on pdf.
3. Generate cards via csv for nuernberg and bavaria. Find bulk import files here: (administration/resources/cards)
4. Import csv from freinet (bavaria)
5. Test csv export for nuernberg
6. Please also activate these cards to check if everything works with the qrCodes and deeplinks



### Resolved issues



Fixes: #1666.
